elastic/kibana
# [Index Management] Support rollover of datastreams from component templates linked to managed/packaged index templates

Closes https://github.com/elastic/kibana/issues/184095

## Summary
This PR enhances the work initially done with https://github.com/elastic/kibana/pull/135845 in order to allow rolling over of related datastreams not only for `component templates` managed by `fleet` but also for those who are referenced by an `index template`  which is part of a package and are also managed.

The UX has remain intact as it was initially designed, the usage of a single checkbox was suggested in the linked issue but its something that perhaps needs to be discussed with a designer involved and can be solved in a separate PR.

##### How to test

* Start up kibana and es
* Add `Elastic Defend` integration to your local installation (using the [following url](http://localhost:5601/dmc/app/integrations/detail/endpoint-8.15.0/overview))
* Navigate to Stack Management -> Index Management -> Component Templates
* Open embeded console and run the commands listed at `Create test datastream`
* Open details for either [logs-elastic_agent.cloud_defend@custom](http://localhost:5601/dmc/app/management/data/index_management/component_templates/logs-elastic_agent.cloud_defend%40custom) or [logs-elastic_agent.cloud_defend@package](http://localhost:5601/dmc/app/management/data/index_management/component_templates/logs-elastic_agent.cloud_defend%40package)
* Verify that:
  * Going thorugh the flow and saving the component template without making any changes and save it. Make sure it doesnt show the rollover mappings modal
  * Open the embded console and run the command listed at `Modify mappings of component template`
  * Edit the same component template you opened before and go through the flow and save the component template without making any changes. Verify that the rollover modal shows up and upon clicking `Apply now and rollover` CTA next time we go through the same flow, it doesnt show the rollover modal.
  * Stopping es/klibana and starting it again, re-adding the integration, running the `Trigger rollover datastreams modal in creation flow` commands in console, creating a new component template with the name of `logs-elastic_agent.cloud_defend@custom` triggers the rollover datastreams modal.

<details>
<summary>Create test datastream</summary>

```
# Creates a component template for mappings
PUT _component_template/my-mappings
{
  "template": {
    "mappings": {
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "date_optional_time||epoch_millis"
        },
        "message": {
          "type": "wildcard"
        }
      }
    }
  },
  "_meta": {
    "description": "Mappings for @timestamp and message fields",
    "my-custom-meta-field": "More arbitrary metadata"
  }
}

PUT _index_template/my-index-template
{
  "index_patterns": ["logs-elastic_agent.cloud_defend*"],
  "data_stream": { },
  "composed_of": [ "my-mappings" ],
  "priority": 500,
  "_meta": {
    "description": "Template for my time series data",
    "my-custom-meta-field": "More arbitrary metadata"
  }
}

POST logs-elastic_agent.cloud_defend-1/_doc
{
  "@timestamp": "2099-05-06T16:21:15.000Z",
  "message": "192.0.2.42 - - [06/May/2099:16:21:15 +0000] \"GET /images/bg.jpg HTTP/1.0\" 200 24736"
}

PUT _component_template/logs-elastic_agent.cloud_defend@custom
{
  "version": 1,
  "template": {}
}
```
</details>

<details>
<summary>Modify mappings of component template</summary>

```
PUT _component_template/my-mappings
{
  "template": {
    "mappings": {
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "date_optional_time||epoch_millis"
        },
        "message": {
          "type": "text"
        }
      }
    }
  },
  "_meta": {
    "description": "Mappings for @timestamp and message fields",
    "my-custom-meta-field": "More arbitrary metadata"
  }
}
```
</details>

<details>
<summary>Trigger rollover datastreams modal in creation flow</summary>

```
# Creates a component template for mappings
PUT _component_template/my-mappings
{
  "template": {
    "mappings": {
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "date_optional_time||epoch_millis"
        },
        "message": {
          "type": "wildcard"
        }
      }
    }
  },
  "_meta": {
    "description": "Mappings for @timestamp and message fields",
    "my-custom-meta-field": "More arbitrary metadata"
  }
}

PUT _index_template/my-index-template
{
  "index_patterns": ["logs-elastic_agent.cloud_defend*"],
  "data_stream": { },
  "composed_of": [ "my-mappings" ],
  "priority": 500,
  "_meta": {
    "description": "Template for my time series data",
    "my-custom-meta-field": "More arbitrary metadata"
  }
}

POST logs-elastic_agent.cloud_defend-1/_doc
{
  "@timestamp": "2099-05-06T16:21:15.000Z",
  "message": "192.0.2.42 - - [06/May/2099:16:21:15 +0000] \"GET /images/bg.jpg HTTP/1.0\" 200 24736"
}

PUT _component_template/my-mappings
{
  "template": {
    "mappings": {
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "date_optional_time||epoch_millis"
        },
        "message": {
          "type": "text"
        }
      }
    }
  },
  "_meta": {
    "description": "Mappings for @timestamp and message fields",
    "my-custom-meta-field": "More arbitrary metadata"
  }
}
```
</details>

<details>
<summary>Screenshots</summary>

<img width="845" alt="Screenshot 2024-07-08 at 11 42 46" src="https://github.com/elastic/kibana/assets/1191206/33e3f6f2-7d13-4d30-9218-f692933228ba">

<img width="886" alt="Screenshot 2024-07-08 at 13 00 16" src="https://github.com/elastic/kibana/assets/1191206/e33ab56c-afbd-4e9b-9d77-0c6490a319a7">


</details>
